### PR TITLE
Feature/hangman piustonen

### PR DESCRIPTION
**Pull request by piustonen**
new adding

### CODE REVIEW

File: hangman.py [hunk 1]
```diff
@@ -1,145 +1,168 @@
...
 
 def is_word_guessed(secret_word, letters_guessed):
-    '''
-    secret_word: string, the word the user is guessing; assumes all letters are
-      lowercase
-    letters_guessed: list (of letters), which letters have been guessed so far;
-      assumes that all letters are lowercase
-    returns: boolean, True if all the letters of secret_word are in letters_guessed;
-      False otherwise
-    '''
```
Comment (Hozhyi) [inline hangman.py]: I believe these comments were quiet valuable, why you removed them?

File: hangman.py [hunk 2]
```diff
@@ -1,145 +1,168 @@
...
-    Depending on the size of the word list, this function may
-    take a while to finish.
-    """
     print("Loading word list from file...")
-    # inFile: file
     inFile = open(WORDLIST_FILENAME, 'r')
-    # line: string
     line = inFile.readline()
-    # wordlist: list of strings
     wordlist = line.split()
```
Comment (Hozhyi) [inline hangman.py]: Not all comments is worth to remove. Only those which won't will be needed in future

File: hangman.py [hunk 3]
```diff
@@ -1,145 +1,168 @@
...
+    counter = 0
+    flag = False
+    for letters in secret_word:
+        for f in letters_guessed:
+            if letters == f:
+                flag = True
+                break
+        if not flag:
+            counter += 1
+        flag = False
```
Comment (Hozhyi) [inline hangman.py]: All this block can be simplified to `if letter in letters_guessed`

File: hangman.py [hunk 4]
```diff
@@ -1,145 +1,168 @@
...
-      which letters in secret_word have been guessed so far.
-    '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
+    s = ""
+    for i in range(len(secret_word)):
+        if secret_word[i] in letters_guessed:
+            s += secret_word[i]
+        else:
+            s += "_ "
```
Comment (Hozhyi) [inline hangman.py]: Try to avoid string concatenating in loop

File: hangman.py [hunk 5]
```diff
@@ -1,145 +1,168 @@
...
+
+    secret_word = s
+    return secret_word
 
+    pass
```
Comment (Hozhyi) [inline hangman.py]: Why it is needed here?

File: hangman.py [hunk 6]
```diff
@@ -1,145 +1,168 @@
...
-    pass
+    s = ""
+    for i in range(len(secret_word)):
+        if secret_word[i] in letters_guessed:
+            s += secret_word[i]
+        else:
+            s += "_ "
+
+    secret_word = s
+    return secret_word
```
Comment (Hozhyi) [inline hangman.py]: Effectively you could just `return s`

File: hangman.py [hunk 7]
```diff
@@ -1,145 +1,168 @@
...
 def get_guessed_word(secret_word, letters_guessed):
-    '''
-    secret_word: string, the word the user is guessing
-    letters_guessed: list (of letters), which letters have been guessed so far
-    returns: string, comprised of letters, underscores (_), and spaces that represents
-      which letters in secret_word have been guessed so far.
-    '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
+    s = ""
```
Comment (Hozhyi) [inline hangman.py]: Please - avoid 1 letter variables

File: hangman.py [hunk 8]
```diff
@@ -1,145 +1,168 @@
...
-    letters_guessed: list (of letters), which letters have been guessed so far
-    returns: string (of letters), comprised of letters that represents which letters have not
-      yet been guessed.
-    '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
-    
-    
+    if len(letters_guessed) == 0:
+        import string
```
Comment (Hozhyi) [inline hangman.py]: Why you decided to move that import here?

File: hangman.py [hunk 9]
```diff
@@ -1,145 +1,168 @@
...
+        x = list(string.ascii_lowercase)
+        print(''.join(x))
```
Comment (Hozhyi) [inline hangman.py]: `string.ascii_lowercase` - is already a string. Why you decided convert it to list, and than split it back to string?

File: hangman.py [hunk 10]
```diff
@@ -1,145 +1,168 @@
...
+    else:
+        import string
```
Comment (Hozhyi) [inline hangman.py]: You could have this `import` only once in the top of file

File: hangman.py [hunk 11]
```diff
@@ -1,145 +1,168 @@
...
+        z = list(string.ascii_lowercase)
+        for elements in letters_guessed:
+            del z[z.index(elements.lower())]
+            y = ''.join(z)  # ''
+        print(y)
```
Comment (Hozhyi) [inline hangman.py]: It shouldn't print it, it only should return it

File: hangman.py [hunk 12]
```diff
@@ -1,145 +1,168 @@
...
-    pass
-    
-    
+    if len(letters_guessed) == 0:
+        import string
+        x = list(string.ascii_lowercase)
+        print(''.join(x))
+    else:
+        import string
+        z = list(string.ascii_lowercase)
```
Comment (Hozhyi) [inline hangman.py]: Effectively - you could just leave that scenario, as it will include the previous one

File: hangman.py [hunk 13]
```diff
@@ -1,145 +1,168 @@
...
+        for elements in letters_guessed:
```
Comment (Hozhyi) [inline hangman.py]: It's not elements, it's a one element. And that's not just element - it's some symbol

File: hangman.py [hunk 14]
```diff
@@ -1,145 +1,168 @@
...
+            del z[z.index(elements.lower())]
```
Comment (Hozhyi) [inline hangman.py]: That could be a good solution, but this operation is complex and time consuming. Try implementing that without deletion

File: hangman.py [hunk 15]
```diff
@@ -1,145 +1,168 @@
...
+    * The user should receive feedback immediately after each guess
       about whether their guess appears in the computer's word.
 
-    * After each guess, you should display to the user the 
+    * After each guess, you should display to the user the
       partially guessed word so far.
-    
+
     Follows the other limitations detailed in the problem write-up.
     '''
```
Comment (Hozhyi) [inline hangman.py:166]: This function is quite long, it should be smaller (up to 30 lines long)

File: hangman.py [hunk 16]
```diff
@@ -1,145 +1,168 @@
...
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
-
+    letters_guessed = []
```
Comment (Hozhyi) [inline hangman.py]: Do you think that `list` is the best container here?

File: hangman.py [hunk 17]
```diff
@@ -1,145 +1,168 @@
...
+    warnings_point = 3
+    guesses_point = 6
```
Comment (Hozhyi) [inline hangman.py]: I'd add a constants for these 3 and 6

File: hangman.py [hunk 18]
```diff
@@ -1,145 +1,168 @@
...
+    print("Welcome to the game Hangman!\nI am thinking of a word that is", len(secret_word), "letters long")
```
Comment (Hozhyi) [inline hangman.py]: Could you have a separate `print` per line? It would be more readable

File: hangman.py [hunk 19]
```diff
@@ -1,145 +1,168 @@
...
+                print("Oops! That letter is not in my word:", get_guessed_word(secret_word, letters_guessed))
+        if is_word_guessed(secret_word, letters_guessed) == 0:
+            final_points = guesses_point * int(len(secret_word))
+            print("Congratulations, you won!")
+            print("Your total score for this game is:", final_points)
+            guesses_point = -1
+        if guesses_point == 0:
+            print("-------")
+            print("Sorry, you ran out of guessed. The word was", secret_word)
+            break
```
Comment (Hozhyi) [inline hangman.py]: Could you have the game result output to be separate from game iterations loop?

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman.py [hunk 20]
```diff
@@ -60,8 +50,15 @@ def is_word_guessed(secret_word, letters_guessed):
...
       False otherwise
     '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
+    counter = 0
+    flag = False
+    for letters in secret_word:
+        if letters in letters_guessed:
+            flag = True
+            break
```
Comment (Hozhyi) [inline hangman.py]: Effectively you could just write here `return False`, and no need in additional `flag`

File: hangman.py [hunk 21]
```diff
@@ -60,8 +50,15 @@ def is_word_guessed(secret_word, letters_guessed):
...
+    if not flag:
+        counter += 1
```
Comment (Hozhyi) [inline hangman.py]: Seems like you can write this code without this `counter` as it just makes this code more complicated.

File: hangman.py [hunk 22]
```diff
@@ -72,76 +69,91 @@ def get_guessed_word(secret_word, letters_guessed):
...
     '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
-
+    formed_word = ""
+    for i in range(len(secret_word)):
+        if secret_word[i] in letters_guessed:
+            formed_word += secret_word[i]
+        else:
+            formed_word += HIDDEN_LETTER
```
Comment (Hozhyi) [inline hangman.py]: This concatenating of strings in loop will produce additional garbage. Try to avoid that

File: hangman.py [hunk 23]
```diff
@@ -72,76 +69,91 @@ def get_guessed_word(secret_word, letters_guessed):
...
-      about whether their guess appears in the computer's word.
-
-    * After each guess, you should display to the user the 
-      partially guessed word so far.
-    
-    Follows the other limitations detailed in the problem write-up.
-    '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
+    import string
```
Comment (Hozhyi) [inline hangman.py]: Why do you moved that line to here? It worth having it in the top of file

File: hangman.py [hunk 24]
```diff
@@ -72,76 +69,91 @@ def get_guessed_word(secret_word, letters_guessed):
...
+    x = list(string.ascii_lowercase)
+    if len(letters_guessed) == 0:
+        return ''.join(x)
+    else:
+        for symbol in letters_guessed:
+            del x[x.index(symbol.lower())]
+            y = ''.join(x)  # ''
```
Comment (Hozhyi) [inline hangman.py]: Could you avoid writing comments that are not useful?

File: hangman.py [hunk 25]
```diff
@@ -72,76 +69,91 @@ def get_guessed_word(secret_word, letters_guessed):
...
-
-    * After each guess, you should display to the user the 
-      partially guessed word so far.
-    
-    Follows the other limitations detailed in the problem write-up.
-    '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
+    import string
+    x = list(string.ascii_lowercase)
```
Comment (Hozhyi) [inline hangman.py]: 1 letter variable

File: hangman.py [hunk 26]
```diff
@@ -72,76 +69,91 @@ def get_guessed_word(secret_word, letters_guessed):
...
+    if len(letters_guessed) == 0:
+        return ''.join(x)
+    else:
+        for symbol in letters_guessed:
+            del x[x.index(symbol.lower())]
```
Comment (Hozhyi) [inline hangman.py]: This deleting from list is not as easy operation. Could you implement different approach? E.g. adding all needed letters to list

File: hangman.py [hunk 27]
```diff
@@ -72,76 +69,91 @@ def get_guessed_word(secret_word, letters_guessed):
...
+    list_word = show_possible_matches(get_guessed_word(secret_word, letters_guessed))
+    if len(list_word) < 30:
 
+        print("Possible word matches are:", " ".join(list_word))
+    else:
+        random_words = []
+        for i in range(30):
+            random_words.append(choose_word(wordlist))
 
+        print("Possible word matches are:", " ".join(random_words))
```
Comment (Hozhyi) [inline hangman.py]: I expected to see it in Danilo's changes only, why you made some changes here? Danilo already merged his changes to develop, so you can re-base your branch on the top of develop, and you will have all code from Danilo

**Review by Hozhyi — CHANGES_REQUESTED** [review]

File: hangman.py [hunk 28]
```diff
@@ -86,17 +94,27 @@ def warnings(value_mistake, warnings_point, guesses_point, secret_word, letters_
...
-        print("Oops! You have no warnings left, so you lose one guess:", get_guessed_word(secret_word, letters_guessed))
+        print("Oops! You have no warnings left, so you lose one guess:"
+              f" {get_guessed_word(secret_word, letters_guessed)}")
 
     return warnings_point, guesses_point
 
 
 def print_hints_warning(list_word):
-    pass
+    if len(list_word) < 30:
```
Comment (MaxShvtsv) [inline hangman.py:115]: Add docs.

File: hangman.py [hunk 29]
```diff
@@ -86,17 +94,27 @@ def warnings(value_mistake, warnings_point, guesses_point, secret_word, letters_
...
+
+        print("Possible word matches are:", " ".join(list_word))
+    else:
+        random_words = []
+        for i in range(30):
+            random_words.append(choose_word(list_word))
+
+    # secret_word while you're doing your own testing)
```
Comment (MaxShvtsv) [inline hangman.py]: What does this comment mean?

File: hangman.py [hunk 30]
```diff
@@ -86,17 +94,27 @@ def warnings(value_mistake, warnings_point, guesses_point, secret_word, letters_
...
+        print("Possible word matches are:", " ".join(random_words))
```
Comment (MaxShvtsv) [inline hangman.py]: Line 137: add comment to `count_star` var

File: hangman.py [hunk 31]
```diff
@@ -153,14 +171,14 @@ def check_letter(letter, warnings_point, guesses_point, secret_word, letters_gue
...
-    if is_word_guessed(secret_word, letters_guessed) == 0:
+    if is_word_guessed(secret_word, letters_guessed):
         final_points = guesses_point * len(set(secret_word))
         print("------")
         print("Congratulations, you won!")
         print("Word:", secret_word)
         print("Your total score for this game is:", final_points)
 
-    elif is_word_guessed(secret_word, letters_guessed) != 0 and guesses_point <= 0:
+    elif not is_word_guessed(secret_word, letters_guessed) and guesses_point <= 0:
```
Comment (MaxShvtsv) [inline hangman.py]: Remove unnecassary condition

File: hangman.py [hunk 32]
```diff
@@ -175,13 +193,13 @@ def hangman(secret_word, with_hints):
     guesses_point = 6
     warnings_point = 3
```
Comment (MaxShvtsv) [inline hangman.py]: Use constants to set value of these vars.

File: hangman.py [hunk 33]
```diff
@@ -194,7 +212,6 @@ def hangman(secret_word, with_hints):
     result_of_game(secret_word, letters_guessed, guesses_point)
 
 
-
 def match_with_gaps(my_word, other_word):
     '''
     my_word: string with _ characters, current guess of secret word
```
Comment (MaxShvtsv) [inline hangman.py:242]: Line 227: you can iterate not only through index but also throuhg elements it iterable objects. Simplify conditions.

File: hangman.py [hunk 34]
```diff
@@ -239,7 +256,7 @@ def user_wants_hints():
```
Comment (MaxShvtsv) [inline hangman.py]: Line 248: don't use True in conditions

File: hangman.py [hunk 35]
```diff
@@ -239,7 +256,7 @@ def user_wants_hints():
 
     while True:
```
Comment (MaxShvtsv) [inline hangman.py]: It is not good to use True in condition of while loop. Change it to boolean var.

File: hangman.py [hunk 36]
```diff
@@ -239,7 +256,7 @@ def user_wants_hints():
...
         type_of_game = input("if you want to play hangman (classic) input 'classic',"
-                             " play hangman with hints input 'hints':").replace(" ", "")
+                             " play hangman with hints input 'hints':").strip()
         if type_of_game == "classic" or type_of_game == "hints":
             if type_of_game == "classic":
                 return False
```
Comment (MaxShvtsv) [inline hangman.py]: Line 264: remove unnecassary conditions

**Review by MaxShvtsv — CHANGES_REQUESTED** [review]
Finished reviewing.
Sorry, but I could not see the whole file, but only changes. In some places I wrote "Line:..." find this line and make changes here.

File: hangman.py [hunk 37]
```diff
@@ -75,8 +81,10 @@ def get_available_letters(letters_guessed):
...
     '''
-    # FILL IN YOUR CODE HERE AND DELETE "pass"
-    pass
+    letters = string.ascii_lowercase
+    for symbol in letters_guessed:
+        letters = letters.replace(symbol, '')
+    return letters
 
 
 def warnings(value_mistake, warnings_point, guesses_point, secret_word, letters_guessed):
```
Comment (MaxShvtsv) [inline hangman.py:90]: Add docs about params  and functions (check also other functions)

File: hangman.py [hunk 38]
```diff
@@ -86,17 +94,30 @@ def warnings(value_mistake, warnings_point, guesses_point, secret_word, letters_
...
+    """
+    function finds words that matches with letters which were chosen by user earlier
+    :param list_word: list of word attached with game
+    """
+    if len(list_word) < 30:
+
+        print("Possible word matches are:", " ".join(list_word))
+    else:
+        random_words = []
+        for i in range(30):
```
Comment (MaxShvtsv) [inline hangman.py]: Use list slicing not loop

File: hangman.py [hunk 39]
```diff
@@ -86,17 +94,30 @@ def warnings(value_mistake, warnings_point, guesses_point, secret_word, letters_
...
+            random_words.append(choose_word(list_word))
+
+        print("Possible word matches are:", " ".join(random_words))
```
Comment (MaxShvtsv) [inline hangman.py]: Line 141: add comment about `count_star` var

File: hangman.py [hunk 40]
```diff
@@ -194,7 +215,6 @@ def hangman(secret_word, with_hints):
     result_of_game(secret_word, letters_guessed, guesses_point)
 
 
-
 def match_with_gaps(my_word, other_word):
     '''
     my_word: string with _ characters, current guess of secret word
```
Comment (MaxShvtsv) [inline hangman.py:242]: Line 231: simplify condition

File: hangman.py [hunk 41]
```diff
@@ -226,7 +246,7 @@ def show_possible_matches(my_word):
     list_with_words = []
     my_word = my_word.replace(" ", "")
     for word in list(wordlist):
```
Comment (MaxShvtsv) [inline hangman.py]: `wordlist` is already list

File: hangman.py [hunk 42]
```diff
@@ -239,17 +259,17 @@ def user_wants_hints():
 
     while True:
```
Comment (MaxShvtsv) [inline hangman.py]: Don't use True in while loop condition

**Review by MaxShvtsv — CHANGES_REQUESTED** [review]

File: hangman.py [hunk 43]
```diff
@@ -225,31 +267,34 @@ def show_possible_matches(my_word):
...
+                         " play hangman with hints input 'hints':").strip()
+    while type_of_game or not type_of_game:
+        if type_of_game == "classic":
+            return False
+        elif type_of_game == "hints":
+            return True
         else:
             print("Oops! Your input is not correct")
+            type_of_game = input("if you want to play hangman (classic) input 'classic',"
+                                 " play hangman with hints input 'hints':").strip()
```
Comment (MaxShvtsv) [inline hangman.py]: Lines of input are repeating. It is not good.

**Review by MaxShvtsv — CHANGES_REQUESTED** [review]

**Review by MaxShvtsv — APPROVED** [review]

**Pull request merged**